src/cyclefold/gadgets.rs

```diff
@@ -0,0 +1,614 @@
+  {
+    pub fn alloc<CS>(
+      mut cs: CS,
+      coords: Option<(E2::Base, E2::Base, bool)>,
+      limb_width: usize,
+      n_limbs: usize,
+    ) -> Result<Self, SynthesisError>
+    where
+      CS: ConstraintSystem<<E1 as Engine>::Base>,
+    {
+      let x = BigNat::alloc_from_nat(
+        cs.namespace(|| "x"),
+        || {
+          Ok(f_to_nat(
+            &coords.ok_or(SynthesisError::AssignmentMissing)?.0,
+          ))
+        },
+        limb_width,
+        n_limbs,
+      )?;
+
+      let y = BigNat::alloc_from_nat(
+        cs.namespace(|| "y"),
+        || {
+          Ok(f_to_nat(
+            &coords.ok_or(SynthesisError::AssignmentMissing)?.1,
+          ))
+        },
+        limb_width,
+        n_limbs,
+      )?;
+
+      let is_infinity = AllocatedNum::alloc(cs.namespace(|| "is_infinity"), || {
+        Ok(if coords.map_or(true, |c| c.2) {
+          E1::Base::ONE
+        } else {
+          E1::Base::ZERO
+        })
+      })?;
+      cs.enforce(
+        || "is_infinity is bit",
+        |lc| lc + is_infinity.get_variable(),
+        |lc| lc + CS::one() - is_infinity.get_variable(),
+        |lc| lc,
+      );
+      Ok(Self {
+        x,
+        y,
+        is_infinity,
+        _p: PhantomData,
+      })
+    }
+
+    pub fn absorb_in_ro<CS>(&self, mut cs: CS, ro: &mut E1::ROCircuit) -> Result<(), SynthesisError>
+    where
+      CS: ConstraintSystem<<E1 as Engine>::Base>,
+    {
+      let x_bn = self
+        .x
+        .as_limbs()
+        .iter()
+        .enumerate()
+        .map(|(i, limb)| {
+          limb.as_allocated_num(cs.namespace(|| format!("convert limb {i} of X_r[0] to num")))
+        })
+        .collect::<Result<Vec<AllocatedNum<E1::Base>>, _>>()?;
+
+      for limb in x_bn {
+        ro.absorb(&limb)
+      }
+
+      let y_bn = self
+        .y
+        .as_limbs()
+        .iter()
+        .enumerate()
+        .map(|(i, limb)| {
+          limb.as_allocated_num(cs.namespace(|| format!("convert limb {i} of X_r[0] to num")))
+        })
+        .collect::<Result<Vec<AllocatedNum<E1::Base>>, _>>()?;
+
+      for limb in y_bn {
+        ro.absorb(&limb)
+      }
+
+      ro.absorb(&self.is_infinity);
+
+      Ok(())
+    }
+
+    pub fn check_on_curve<CS>(
+      &self,
+      mut cs: CS,
+      limb_width: usize,
+      n_limbs: usize,
+    ) -> Result<(), SynthesisError>
+    where
+      CS: ConstraintSystem<<E1 as Engine>::Base>,
+    {
+      let m_bn = alloc_bignat_constant(
+        cs.namespace(|| "alloc m"),
+        &E1::GE::group_params().3,
+        limb_width,
+        n_limbs,
+      )?;
+
+      let A_bn = alloc_bignat_constant(
+        cs.namespace(|| "alloc A"),
+        &f_to_nat(&E1::GE::group_params().0),
+        limb_width,
+        n_limbs,
+      )?;
+
+      let B_bn = alloc_bignat_constant(
+        cs.namespace(|| "alloc B"),
+        &f_to_nat(&E1::GE::group_params().1),
+        limb_width,
+        n_limbs,
+      )?;
```

Comment:
It would be better to allocate these constants once (if they really need to be allocated at all) rather than every time `check_on_curve` is called.